---
title: "Mycoplasma Report"
author: "Alfredo Rago"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
# params: 
#  sample_id : snakemake@input_base_fq
#  mapping_counts : snakemake@input
---

```{r setup, include=FALSE}
library(tidyverse)
library(ggplot2)
library(janitor)
library(here)
print(paste("Running script from ", here()))
```

```{r load_data, include=FALSE}
file_location = here("results/fastq_screen/")
mapping_count_files = list.files(path = file_location, pattern = "*_screen.txt", full.names = T) %>% 
  set_names(list.files(path = file_location, pattern = "*_screen.txt"))

counts_report = map_dfr(mapping_count_files, read_tsv, skip = 1, .id = "read_id") %>%
  add_column(sample_id = str_extract(string = .$read_id, pattern = "A_[0-9]{1,2}"), .after = 1) %>% 
  mutate(., 
         read_id = str_extract(read_id, "A_[0-9]_[1,2]")) %>% 
  clean_names() %>% 
  rename(., number_multiple_hits_multiple_genomes = multiple_hits_multiple_genomes) %>% 
  pivot_longer(
    data = ., 
    cols = number_reads_processed:percent_multiple_hits_multiple_genomes, 
    names_pattern = "([^_]*)_(.*)", names_to = c("percent_number", "measurement"), values_to = "counts"
  )

```

Several cultures in the lab showed contamination from Mycoplasma, which would render the samples unusable.
I check for contamination in the RNAseq data by mappint the reads to the Human and Mycoplasma genomes, as well as a set of reference organisms and contaminants.
The mapping is done via the program fastq_screen, which is essentially a wrapper for bowtie2 alignment.

## Raw read counts to Human and contaminants


```{r mapped_counts,  echo=FALSE}
filter(.data = counts_report, 
       genome %in% c("Human", "Mycoplasma", "Rat", "Ecoli", "rRNA", "MT", "Adapters"),
       # measurement %in% c("one_hit_one_genome", "multiple_hits_one_genome"),
       percent_number == 'number') %>% 
  ggplot(data = ., 
       mapping = aes(
         x = counts, 
         y = sample_id, 
         col = genome,
         group = read_id)
       ) +
  geom_point() +
  geom_hline(yintercept = 0.1) +
  scale_color_brewer(type = 'qual') +
  facet_wrap(. ~ measurement) +
  xlab("Number of reads mapped") +
  ylab(label = "Sample ID") +
  ggtitle(label = "Number of reads mapped to Human and contaminant genomes") +
  theme_minimal()

```

We have overall excellent mapping to human samples, and almost no mapping to other organisms. Matches in the "Multiple genomes" categories are expected, since they match orthologs between species. Unsurprisingly, we find a substantial amount of e.g. Rat-like RNA. 

The only reads that map to Mycoplasma are multi-hit single genome. If these are really from Mycoplasma, then they are most likely caused by the expression of paralogs.


## Proportions of mapped reads to Human and contaminant genomes

```{r mapped_percent,  echo=FALSE}
filter(.data = counts_report, 
       genome %in% c("Human", "Mycoplasma", "Rat", "Ecoli", "rRNA", "MT", "Adapters"),
       # measurement %in% c("one_hit_one_genome", "multiple_hits_one_genome"),
       percent_number == 'percent') %>% 
  ggplot(data = ., 
       mapping = aes(
         x = counts, 
         y = sample_id, 
         col = genome,
         group = read_id)
       ) +
  geom_point() +
  geom_hline(yintercept = 0.1) +
  scale_color_brewer(type = 'qual') +
  facet_wrap(. ~ measurement) +
  xlab(label = "Percent Reads Mapped") +
  ylab(label = "Sample ID") +
  ggtitle(label = "Percentage of reads mapped to Human and contaminant genomes") +
  theme_minimal()
```

Matches to human genome are spread roughly as 50% single hit and 50% multiple hit. Half of the multi-hit reads are also found in other genomes (Ribosomal, Rat and other vertebrates).
Samples 58 and 55 show a small amount of Mycoplasma matching reads, and have a higher proportion of reads that do not map to the Human genome. These may be contaminated and should be removed from our analysis.
